packages/cheqd-sdk/src/dids/CheqdDidResolver.ts

```diff
@@ -0,0 +1,202 @@
+      if (!parsedDid) {
+        throw new Error('Invalid DID')
+      }
+
+      switch (did) {
+        case did.match(cheqdDidRegex)?.input:
+          return await this.Resolve(agentContext, parsedDid.did)
+        case did.match(cheqdDidVersionRegex)?.input: {
+          const version = did.match(cheqdDidVersionRegex)![7]
+          return await this.Resolve(agentContext, parsedDid.did, version)
+        }
+        case did.match(cheqdDidVersionsRegex)?.input:
+          return await this.ResolveAllDidDocVersions(agentContext, parsedDid)
+        case did.match(cheqdDidMetadataRegex)?.input:
+          return await this.DereferenceCollectionResources(agentContext, parsedDid)
+        case did.match(cheqdResourceMetadataRegex)?.input:
+          return await this.DereferenceResourceMetadata(agentContext, parsedDid)
+        default:
+          return {
+            didDocument: null,
+            didDocumentMetadata,
+            didResolutionMetadata: {
+              error: 'Invalid request',
+              message: `Unsupported did Url: '${did}'`,
+            },
+          }
+      }
+    } catch (error) {
+      return {
+        didDocument: null,
+        didDocumentMetadata,
+        didResolutionMetadata: {
+          error: 'notFound',
+          message: `resolver_error: Unable to resolve did '${did}': ${error}`,
+        },
+      }
+    }
+  }
+
+  public async resolveResource(agentContext: AgentContext, did: string): Promise<any> {
+    const cheqdSdkLedgerService = agentContext.dependencyManager.resolve(CheqdSdkLedgerService)
+    try {
+      const parsedDid = parseCheqdDid(did)
+      if (!parsedDid) {
+        throw new Error('Invalid DID')
+      }
+
+      const { id, params, network } = parsedDid
+      await cheqdSdkLedgerService.connect({ network })
+
+      let resourceId: string
+      if (did.match(cheqdResourceRegex)?.input) {
+        resourceId = did.split('/')[2]
+      } else if (params && params.resourceName && params.resourceType) {
+        let resources = (await cheqdSdkLedgerService.resolveCollectionResources(id)).resources
+        resources = resources.filter(
+          (resource) => resource.name == params.resourceName && resource.resourceType == params.resourceType
+        )
+        if (!resources.length) {
+          throw new Error(`No resources found`)
+        }
+
+        let resource: Metadata | undefined
+        if (params.version) {
+          resource = resources.find((resource) => resource.version == params.version)
+        } else {
+          const date = params.time ? new Date(params.time) : new Date()
+          // find the resourceId to the created time
+          resources.sort(function (a, b) {
+            const distancea = Math.abs(date.getTime() - a.created!.getTime())
+            const distanceb = Math.abs(date.getTime() - b.created!.getTime())
+            return distancea - distanceb
+          })
+          resource = resources[0]
+        }
```

Comment:
Maybe we can extract some of these things into util methods so it'll become easier to understand the flow of what's happening here, without needing to understand the implementation.

E.g.:

```ts
resources = filterResourcesByNameAndType(resources)

if (params.version) {
  resource = findResourceByVersion(resources)
} else {
  const date = params.time ? new Date(params.time) : new Date()
  resources = sortResourcesByDateDescending(date)

  resource = resources[0]
}
```